### ISSUE TITLE
DeveloperGuide: Update Appendix A.1 Storage component

### ISSUE THREAD

**Issue by willxujun** [se-edu/addressbook-level4#708]
Developer guide's appendix A.1 storage component has a solution which links to a PR. This PR fails Travis CI build, because XmlAddressBookStorage does not implement abstract method backupAddressBook. Is the API method supposed to be written in Storage interface instead of AddressBookStorage interface? Because the latter has no real reason for a new backup method, since the logic implemented in StorageManager class does not require it.

**Comment by damithc** [comment]
@yamgent any thoughts?

**Comment by Zhiyuan-Amos** [comment]
I think you're right. ^^ 

`XmlAddressBookStorage` already has `saveAddressBook(ReadOnlyAddressBook addressBook, String filePath)` to specify saving to a different location. 
As such, I think the API method should be written in `Storage` interface, and `StorageManager` implements it.

**Comment by yamgent** [comment]
> Is the API method supposed to be written in Storage interface instead of AddressBookStorage interface? Because the latter has no real reason for a new backup method, since the logic implemented in StorageManager class does not require it.

The method is supposed to be placed in `AddressBookStorage`, because `backupAddressBook()` is a method specific to address books, not to any storage in general.

Having a `backupAddressBook()` allows you to generate a backup file without having the caller to pass in the filepath of the backup file. Whether that is useful is debatable (depends on how you bring the project further), but at least that was the intention of the method.

**Comment by yamgent** [comment]
> because XmlAddressBookStorage does not implement abstract method backupAddressBook.

And yes you are right, I did not implement `backupAddressBook` for `XmlAddressBookStorage`. My bad. It should also saves to a backup file for the xml version (e.g. `abcd-**backup.xml**`).

**Comment by Zhiyuan-Amos** [comment]
> The method is supposed to be placed in AddressBookStorage, because backupAddressBook() is a method specific to address books, not to any storage in general.

This is arguable since `Storage` has methods like `readAddressBook()`, `saveAddressBook(ReadOnlyAddressBook)`, so having the concept of a `address book` in `Storage` seems ok (Also, `Storage extends AddressBookStorage`). 

But I think we can leave this as that since we didn't have time to tidy up the `Storage` component :P  There are definitely better ways to improve the design.

**Comment by Zhiyuan-Amos** [comment]
After having a clearer view of what this is supposed to do from Wang Leng, I think for this issue we should do something like: Instead of having a `backupAddressBook()` method, we can have a `backupData()` to be implemented in the `Storage` interface. Since other components access the Storage component of the program through `Storage`, it's necessary that `Storage` interface exposes a method to backup data. 

> Having a backupAddressBook() allows you to generate a backup file without having the caller to pass in the filepath of the backup file. Whether that is useful is debatable (depends on how you bring the project further), but at least that was the intention of the method.

I agree that this is debatable. The `backupData()` method in `StorageManager` can look something like:
```
public void backupData() {
    abStorage.saveAddressBook(abStorage.readAddressBook(), BACKUP_FILEPATH);
}
```

or

```
public void backupData() {
    abStorage.backupData();
}
```

I think both are fine as of the current design.

**Comment by Zhiyuan-Amos** [comment]
Should be resolved as #594 is updated.
